iceoryx_hoofs/test/moduletests/test_static_lifetime_guard.cpp

```diff
@@ -0,0 +1,224 @@
+// Copyright (c) 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_hoofs/design_pattern/static_lifetime_guard.hpp"
+
+#include "test.hpp"
+#include <iostream>
+
+namespace
+{
+using namespace ::testing;
+
+struct Foo
+{
+    Foo()
+        : id(++instancesCreated)
+    {
+        ++ctorCalled;
+    }
+
+    ~Foo()
+    {
+        ++dtorCalled;
+    }
+
+    static void reset()
+    {
+        ctorCalled = 0;
+        dtorCalled = 0;
+    }
+
+    static uint32_t ctorCalled;
+    static uint32_t dtorCalled;
+    static uint32_t instancesCreated;
+
+    uint32_t id;
+};
+
+uint32_t Foo::ctorCalled{0};
+uint32_t Foo::dtorCalled{0};
+uint32_t Foo::instancesCreated{0};
+
+using Guard = iox::design_pattern::StaticLifetimeGuard<Foo>;
+
+// the first call to instance() creates a static instance,
+// g_instance is guarded once implicitly
+Foo& g_instance = Guard::instance();
```

Comment:
Please move this to a function, e.g.
```cpp
Foo& gInstance() {
    static Foo& f = Guard::instance();
    return f;
}
```
or always use `Guard::instance();` directly.

This makes it possible to have this test
```cpp
{
    Guard g;
}
EXPECT_THAT(...);
```
No ctor/dtor should be called when this is the first test.

Reply:
> or always use Guard::instance();

Would not be in line with the intention. `g_instance` is set once and never changes while `Guard::instance` may (with counter manipulation).

Reply:
An alternative would of course be make `Foo` a non type template and use `Foo<1>`, `<Foo<2>`, etc. Then it also possible to create the test I proposed.